### PR TITLE
Tweak sender names.

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
PTAL @michaelkirk 

* Tweaks "sender names" in group conversations.
* Tweaks footer layout.

### CODE REVIEW

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 1]
```diff
@@ -970,13 +1059,20 @@ - (CGSize)measureSize
 
     CGSize cellSize = CGSizeZero;
 
+    // TODO: Reflect "sender name" and "footer" layout.
+    //    shouldFooterOverlayMedia = self.canFooterOverlayMedia;
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: This is mostly done but will be completed in a subsequent PR.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 2]
```diff
@@ -359,72 +353,124 @@ - (void)configureViews
...
     }
 
-    OWSMessageFooterView *footerView = self.footerView;
-    [footerView configureWithConversationViewItem:self.viewItem];
-    if (textStackView) {
-        // Display footer below text.
-        [textStackView addArrangedSubview:self.footerView];
-        [self.footerView setHasShadows:NO viewItem:self.viewItem];
-    } else if (bodyMediaView) {
+    if (self.viewItem.shouldHideFooter) {
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: TIOLI: Maybe a little clearer to consolidate the logic that affects whether the footer overlays media here:

```
BOOL shouldFooterOverlayMedia = bodyMediaView && self.canFooterOverlayMedia &&  !self.viewItem.shouldHideFooter;
if (shouldFooterOverlayMedia) {
...
} else {
  // Display footer at bottom of message bubble.
  ...
}
```

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 3]
```diff
@@ -359,72 +353,124 @@ - (void)configureViews
...
+        // Do nothing.
+    } else if (shouldFooterOverlayMedia) {
+        OWSAssert(bodyMediaView);
         // Display footer over media.
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: nit: the code is pretty literate, not sure these comments add anything.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 4]
```diff
@@ -359,72 +353,124 @@ - (void)configureViews
...
+            return YES;
+        case OWSMessageCellType_Audio:
+        case OWSMessageCellType_GenericAttachment:
+        case OWSMessageCellType_DownloadingAttachment:
+        case OWSMessageCellType_ContactShare:
+            return NO;
+    }
+}
+
+- (void)insertTextViewsIntoStackViewIfNecessary:(NSArray<UIView *> *)textViews
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: TIOLI naming nit: `insertAnyTextViewsIntoStackView:` to succinctly surface what is "necessary" for them to be added.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 5]
```diff
@@ -359,72 +353,124 @@ - (void)configureViews
...
+                                 [self autoSetDimension:ALDimensionWidth toSize:bubbleSize.width],
+                             ]];
+                         }];
+
+    [self updateBubbleColorWithBodyMediaView:bodyMediaView];
+}
+
+- (void)updateBubbleColorWithBodyMediaView:(nullable UIView *)bodyMediaView
+{
+    BOOL hasOnlyBodyMediaView = NO;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: TIOLI naming nit: If found the current name confusing. Would something like `isMediaWithNoCaption` capture the same concept?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 6]
```diff
@@ -359,72 +353,124 @@ - (void)configureViews
...
+        case OWSMessageCellType_ContactShare:
+            break;
+        case OWSMessageCellType_StillImage:
+        case OWSMessageCellType_AnimatedImage:
+        case OWSMessageCellType_Audio:
+        case OWSMessageCellType_Video:
+            hasOnlyBodyMediaView = (bodyMediaView && self.stackView.subviews.count == 1);
+            break;
+    }
+    if ([self.viewItem.interaction isKindOfClass:[TSMessage class]] && !hasOnlyBodyMediaView) {
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: In what case would self.viewItem.interaction not be a TSMessage? If none, a guard pattern/owsFail would be clearer.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 7]
```diff
@@ -961,6 +1022,34 @@ - (CGSize)quotedMessageSize
...
+    if (!self.shouldShowSenderName) {
+        return CGSizeZero;
+    }
+
+    CGFloat hMargins = self.conversationStyle.textInsetHorizontal * 2;
+    const int maxTextWidth = (int)floor(self.conversationStyle.maxMessageWidth - hMargins);
+    [self configureSenderNameLabel];
+    CGSize result = CGSizeCeil([self.senderNameLabel sizeThatFits:CGSizeMake(maxTextWidth, CGFLOAT_MAX)]);
+
+    BOOL hasSeparateTextStackView = (self.isQuotedReply || bodyMediaSize.width > 0 || bodyMediaSize.height > 0);
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: TIOLI: It's not obvious what bodyMediaSize has to do with whether or not there is a textStackView. Is that just a check to see if there **is** media? e.g. could we pull out a self.hasMedia getter for clarity?

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m [hunk 8]
```diff
@@ -970,13 +1059,20 @@ - (CGSize)measureSize
 
     CGSize cellSize = CGSizeZero;
 
+    // TODO: Reflect "sender name" and "footer" layout.
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Is this still TODO? It seems like it's addressed here.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageFooterView.h [hunk 9]
```diff
@@ -8,12 +8,10 @@ NS_ASSUME_NONNULL_BEGIN
 
 @interface OWSMessageFooterView : UIStackView
 
-- (void)configureWithConversationViewItem:(ConversationViewItem *)viewItem;
+- (void)configureWithConversationViewItem:(ConversationViewItem *)viewItem hasShadows:(BOOL)hasShadows;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/Cells/OWSMessageFooterView.h]: :+1:

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 10]
```diff
@@ -4875,27 +4878,51 @@ - (void)reloadViewItems
...
+
+            shouldShowSenderAvatar = viewItem.isGroupThread && !isCollapsed;
+        }
+        lastInteractionType = interactionType;
+
+        viewItem.shouldShowSenderAvatar = shouldShowSenderAvatar;
+        viewItem.shouldHideFooter = shouldHideFooter;
+    }
+
+    // Iterate again in forward order.
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: PERF TODO: not sure if this is a hotspot, but it seems like we could combine these iterations by keeping a reference to the previousViewItem.

It's OK for this PR though.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.h [hunk 11]
```diff
@@ -55,10 +55,9 @@ NSString *NSStringForOWSMessageCellType(OWSMessageCellType cellType);
 @property (nonatomic, readonly) BOOL hasQuotedText;
 
 @property (nonatomic) BOOL shouldShowDate;
-// TODO: Consider renaming to shouldHideFooter.
-@property (nonatomic) BOOL shouldHideRecipientStatus;
-// Used to suppress "group sender" avatars.
-@property (nonatomic) BOOL shouldHideAvatar;
+@property (nonatomic) BOOL shouldShowSenderAvatar;
```
Comment (michaelkirk-signal) [inline Signal/src/ViewControllers/ConversationView/ConversationViewItem.h]: It seems a little strange that some of these BOOL's are phrased as Show and others as Hide, but no biggie.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 8 above)
Comment (michaelkirk-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: > This is mostly done but will be completed in a subsequent PR.

Sorry - I started my review before your comment. Ok!

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 2 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: It's ternary:

* if `shouldHideFooter` is true, we don't want to show the footer.
* if `shouldFooterOverlayMedia`, we show footer over media thumbnail.
* else show footer at bottom of the message bubble.

So we can't combine them.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 3 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Removed.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 4 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Done.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 5 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: I've reworked this in a subsequent PR - let's discuss there.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 6 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: Removed.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m (hunk 7 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSMessageBubbleView.m]: I completely agree, so I rewrote this in a subsequent PR.

File: Signal/src/ViewControllers/ConversationView/ConversationViewItem.h (hunk 11 above)
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewItem.h]: The intention was to frame these in such a way that the common case was NO.

**Pull request merged**